agent/connect/ca/provider_aws.go

```diff
@@ -0,0 +1,700 @@
+	}
+
+	if a.isPrimary {
+		// Just use the cert as a root
+		a.rootPEM = *output.Certificate
+	} else {
+		a.intermediatePEM = *output.Certificate
+		// TODO(banks) support user-supplied CA being a Subordinate even in the
+		// primary DC. For now this assumes there is only one cert in the chain
+		if output.CertificateChain == nil {
+			return fmt.Errorf("Subordinate CA %s returned no chain", a.arn)
+		}
+		a.rootPEM = *output.CertificateChain
+	}
+	return nil
+}
+
+func (a *AWSProvider) signCSRRaw(csr *x509.CertificateRequest, templateARN string, ttl time.Duration) (string, error) {
+	// PEM encode the CSR
+	var pemBuf bytes.Buffer
+	if err := pem.Encode(&pemBuf, &pem.Block{Type: "CERTIFICATE REQUEST", Bytes: csr.Raw}); err != nil {
+		return "", err
+	}
+
+	return a.signCSR(pemBuf.String(), templateARN, ttl)
+}
+
+// pollWait returns how long to wait for the next poll of an async operation. We
+// optimize for times typically seen in the API. This is called _before_ the
+// first poll so we can provide a typical delay since operations are never
+// complete immediately so it's a waste to try.
+func pollWait(attemptsMade int) time.Duration {
+	// Hard code times for now
+	waits := []time.Duration{
+		// Never seen it complete first time with a lower value
+		100 * time.Millisecond,
+		200 * time.Millisecond,
+		500 * time.Millisecond,
+		1 * time.Second,
+		3 * time.Second,
+		5 * time.Second,
+	}
+	maxIdx := len(waits) - 1
+	if attemptsMade > maxIdx {
+		attemptsMade = maxIdx
+	}
+	return waits[attemptsMade]
+}
+
+func (a *AWSProvider) pollLoop(desc string, timeout time.Duration, f func() (bool, string, error)) (string, error) {
+	attemptsMade := 0
+	start := time.Now()
+	wait := pollWait(attemptsMade)
+	for {
+		elapsed := time.Since(start)
+		if elapsed >= timeout {
+			return "", fmt.Errorf("timeout after %s waiting for %s", elapsed, desc)
+		}
+
+		a.logger.Printf("[DEBUG] connect.ca.aws: %s pending"+
+			", waiting %s to check readiness", desc, wait)
+		select {
+		case <-a.stopCh:
+			// Provider discarded
+			a.logger.Print("[WARN] connect.ca.aws: provider instance terminated"+
+				" while waiting for %s.", desc)
+			return "", fmt.Errorf("provider terminated")
+		case <-time.After(wait):
+			// Continue looping...
+		}
+
+		done, out, err := f()
+		if err != nil {
+			return "", err
+		}
+		if done {
+			return out, err
+		}
+
+		attemptsMade++
+		wait = pollWait(attemptsMade)
+	}
+}
+
+func (a *AWSProvider) signCSR(csrPEM string, templateARN string, ttl time.Duration) (string, error) {
+	_, signAlg, err := keyTypeToAlgos(a.config.PrivateKeyType, a.config.PrivateKeyBits)
+	if err != nil {
+		return "", err
+	}
+
+	issueInput := acmpca.IssueCertificateInput{
+		CertificateAuthorityArn: aws.String(a.arn),
+		Csr:                     []byte(csrPEM),
+		SigningAlgorithm:        aws.String(signAlg),
+		TemplateArn:             aws.String(templateARN),
+		Validity: &acmpca.Validity{
+			Value: aws.Int64(int64(ttl.Seconds() / 86400.0)),
```

Comment:
nit: I know the stdlib explicitly does not provide a `time.Day` due to DST issues, but seeing "24 hours" in code may be somewhat more obvious than "86400 seconds" to folks who don't play with clocks all the time.
```
const roughDay = 24 * time.Hour
aws.Int64(int64(ttl / roughDay))
```